AdvancedDLSupport/ImplementationGenerators/Wrappers/DelegateWrapper.cs

```diff
@@ -0,0 +1,229 @@
+                {
+                    // Load this
+                    il.EmitLoadArgument(0);
+                    il.EmitLoadArgument(i);
+                    il.Emit(OpCodes.Call, allocMethod);
+                }
+                else
+                {
+                    il.EmitLoadArgument(i);
+                }
+            }
+        }
+
+        /// <inheritdoc/>
+        public override void EmitEpilogue(ILGenerator il, PipelineWorkUnit<IntrospectiveMethodInfo> workUnit)
+        {
+            // If the return type is a delegate, convert it back into its generic representation
+            var definition = workUnit.Definition;
+            var returnType = definition.ReturnType;
+
+            if (!returnType.IsDelegate())
+            {
+                return;
+            }
+
+            var marshalPointerToDel = typeof(Marshal).GetMethods(BindingFlags.Public | BindingFlags.Static).FirstOrDefault(x =>
+                x.IsGenericMethodDefinition && x.Name == "GetDelegateForFunctionPointer" && x.GetParameters().Length == 1 && x.GetParameters()[0].ParameterType == typeof(IntPtr));
+
+            if (marshalPointerToDel == null)
+            {
+                throw new InvalidOperationException("Marshal.GetDelegateForFunctionPointer<T>(IntPtr) not found");
+            }
+
+            var intPtrEquality = typeof(IntPtr).GetMethod("op_Equality");
+            var intPtrZero = typeof(IntPtr).GetField("Zero");
+
+            var retNullLabel = il.DefineLabel();
+            var intPtrRetVal = il.DeclareLocal(typeof(IntPtr));
+
+            il.Emit(OpCodes.Stloc_0);
+            il.Emit(OpCodes.Ldloc_0);
+
+            il.Emit(OpCodes.Ldsfld, intPtrZero);
+            il.Emit(OpCodes.Call, intPtrEquality);
+
+            il.Emit(OpCodes.Brtrue_S, retNullLabel);
+
+            il.Emit(OpCodes.Ldloc_0);
+
+            marshalPointerToDel = marshalPointerToDel.MakeGenericMethod(workUnit.Definition.ReturnType);
+
+            il.Emit(OpCodes.Call, marshalPointerToDel);
+            il.Emit(OpCodes.Ret);
+
+            il.MarkLabel(retNullLabel);
+
+            il.Emit(OpCodes.Ldnull);
+        }
+
+        /// <summary>
+        /// Gets the delegate lifetime that the parameter with the given attributes should be marshalled with.
+        /// </summary>
+        /// <param name="customAttributes">The custom attributes applied to the parameter.</param>
+        /// <returns>The delegate lifetime.</returns>
+        [Pure]
+        private DelegateLifetime GetParameterDelegateLifetime([NotNull, ItemNotNull] IEnumerable<CustomAttributeData> customAttributes)
+        {
+            var lifetimeAttribute = customAttributes.FirstOrDefault
+            (
+                a =>
+                    a.AttributeType == typeof(DelegateLifetimeAttribute)
+            );
+
+            if (lifetimeAttribute is null)
+            {
+                // Default to keeping delegates alive
+                return DelegateLifetime.Persistent;
```

Comment:
Should we default to this? Not in line with CLR P/Invoke, but also very handy. Discuss?

Reply:
I agree with this change. No, it's not inline with P/Invoke, but the advantages outweigh the disadvantages - it will save having to document cases like these and iirc delegates are only 32 bytes in size. Granted, if a method using a delegate parameter is called lots the memory could get very full of stale cache; to circumvent this we could introduce an `UntilNextCall` option which will cache the delegate until the __same method__ is called again.

Reply:
Calling the same method with the same parameter should not add to the cache. `UntilNextCall` is interesting, but really, in what case would that not be solved by a user-managed lifetime and also be relevant? Native code is rarely as frivolous with callbacks as C# is.

Reply:
In my opinion it's a way of removing the need for user-managed lifetime. If we don't make persistent or next-call default, people will be confused and we'll probably get question issues asking why delegates don't work.

UntilNextCall will suffice in nearly all cases, and removes the need for the end user to worry about delegate lifetimes. I'd really like this to be an under-the-hood thing rather than something the user needs to configure.

Reply:
Obviously it's up to @jvbsl whether this is in the scope for this PR, but it'll be a great addition that'll solve nearly all problems. In my mind the user should only use the DelegateLifetimeAttribute if absolutely necessary.

Reply:
Well problem with "UntilNextCall" would be(as someone else mentioned on discord) that it can be dependant on other parameters of the call, perhaps one would need a, DependantOnParameters = nameof(param1). Or similar?

Reply:
I'm on about a simple until next call caching. No fancy nameofs or anything as user-managed lifetimes can be used if you have a fancy situation.